File: web/src/model/saveFile.ts
import { COLUMNS, RESERVE_SIZE } from './card';
import type { Step } from './board';
import { Game } from './game';
import { ruleFor } from './rule';

/*
 * The same text format as the Java version's save files:
 *
 *   level
 *   reserve cards, or 99 once dealt
 *   one line per column, top card first
 *   deal    number          (optional)
 *   time    milliseconds    (optional)
 *   steps   history         (optional)
 *
 * Card values are 0-51, negative (-value-1) when face down; fields are tab-separated.
 */

export function toText(game: Game): string {
  const b = game.board;
  const lines: string[] = [game.rule.level];
  const reserve = b.reserveUsed ? [99, 99, 99] : b.reserve.map((c) => -c - 1);
  lines.push(reserve.map((v) => v + '\t').join(''));
  for (const col of b.columns) lines.push(col.map((s) => (s.up ? s.card : -s.card - 1) + '\t').join(''));
  lines.push(`deal\t${game.dealNumber}`);
  lines.push(`time\t${Math.round(game.elapsedMs)}`);
  const steps = game.history.map((s) =>
    s.kind === 'reserve'
      ? 'r'
      : `m,${s.move.fromColumn},${s.move.fromRow},${s.move.toColumn},${s.count},${s.flipped ? 1 : 0}`,
  );
  lines.push(['steps', ...steps].join('\t'));
  return lines.join('\n') + '\n';
}

export function fromText(text: string): Game {
  const lines = text.split(/\r?\n/);
  const int = (s: string): number => {
    if (!/^-?\d+$/.test(s.trim())) throw new Error('Not a saved game');
    return Number(s);
  };
  const game = new Game(ruleFor(lines[0] ?? ''));
  const b = game.board;

  const reserve = (lines[1] ?? '').split('\t').slice(0, RESERVE_SIZE).map(int);
  if (reserve.length !== RESERVE_SIZE) throw new Error('Not a saved game');
  b.reserve = reserve[0] === 99 ? [] : reserve.map((v) => (v < 0 ? -v - 1 : v));

  for (let c = 0; c < COLUMNS; c++) {
    if (lines[2 + c] === undefined) throw new Error('Not a saved game');
    b.columns[c] = lines[2 + c]
      .split('\t')
      .filter((f) => f.length > 0)
      .map(int)
      .map((v) => (v < 0 ? { card: -v - 1, up: false } : { card: v, up: true }));
  }

  const steps: Step[] = [];
  for (const line of lines.slice(2 + COLUMNS)) {
    const fields = line.split('\t');
    if (fields[0] === 'deal') game.dealNumber = int(fields[1]);
    else if (fields[0] === 'time') game.elapsedMs = int(fields[1]);
    else if (fields[0] === 'steps') {
      for (const f of fields.slice(1)) {
        if (f === 'r') {
          steps.push({ kind: 'reserve' });
        } else {
          const p = f.split(',');
          steps.push({
            kind: 'move',
            move: { fromColumn: int(p[1]), fromRow: int(p[2]), toColumn: int(p[3]) },
            count: int(p[4]),
            flipped: p[5] === '1',
          });
        }
      }
    }
  }
  game.restoreHistory(steps);
  return game;
}

File: src/com/family/solitaire/model/SaveFile.java
package com.family.solitaire.model;

import static com.family.solitaire.model.CardConstants.NCOLS;
import static com.family.solitaire.model.CardConstants.RESERVE_SIZE;

import java.io.BufferedReader;
import java.io.File;
import java.io.FileReader;
import java.io.FileWriter;
import java.io.IOException;
import java.io.Reader;
import java.io.StringReader;
import java.io.StringWriter;
import java.io.Writer;
import java.nio.file.Files;
import java.nio.file.StandardCopyOption;
import java.util.ArrayList;
import java.util.List;

/**
 * Reads and writes games as text.
 * <pre>
 * level
 * reserve cards, or 99 once dealt
 * one line per column, top card first
 * deal    number          (optional, added in 1.4)
 * time    milliseconds    (optional)
 * steps   history         (optional)
 * </pre>
 * Card values are 0-51, negative (-value-1) when face down; fields are
 * tab-separated. Files from older versions have only the first nine lines.
 *
 * @author Aaron Ding
 */
public final class SaveFile {

    private SaveFile() { }

    public static Game load(File file) throws IOException {
        try (Reader reader = new FileReader(file)) {
            return read(reader);
        }
    }

    /** Writes to a temporary file first so a crash never leaves half a save. */
    public static void save(Game game, File file) throws IOException {
        File tmp = new File(file.getPath() + ".tmp");
        try (Writer writer = new FileWriter(tmp)) {
            write(game, writer);
        }
        Files.move(tmp.toPath(), file.toPath(), StandardCopyOption.REPLACE_EXISTING);
    }

    public static String toText(Game game) {
        StringWriter writer = new StringWriter();
        try {
            write(game, writer);
        } catch (IOException e) {
            throw new IllegalStateException(e);
        }
        return writer.toString();
    }

    public static Game fromText(String text) throws IOException {
        return read(new StringReader(text));
    }

    public static void write(Game game, Writer writer) throws IOException {
        Board board = game.getBoard();
        writer.write(game.getLevel());
        writer.write("\n");
        for (Card card : board.getReserve().getCards())
            writer.write((card == null ? "99" : String.valueOf(card.toValue())) + "\t");
        writer.write("\n");
        for (int c=0; c<NCOLS; c++) {
            Column col = board.getColumn(c);
            for (int row=0; row<col.getSize(); row++)
                writer.write(col.getCard(row).toValue() + "\t");
            writer.write("\n");
        }
        writer.write("deal\t" + game.getDealNumber() + "\n");
        writer.write("time\t" + game.getElapsedMillis() + "\n");
        writer.write("steps");
        for (Step step : game.getHistory()) {
            writer.write("\t");
            if (step.isReserveDeal()) {
                writer.write("r");
            } else {
                Move m = step.getMove();
                writer.write("m," + m.fromColumn + "," + m.fromRow + "," + m.toColumn
                        + "," + step.getCount() + "," + (step.flipped() ? 1 : 0));
            }
        }
        writer.write("\n");
    }

    public static Game read(Reader in) throws IOException {
        BufferedReader reader = new BufferedReader(in);
        Game game = new Game();
        Board board = game.getBoard();
        board.clear();
        try {
            Rule rule = Rule.forLevel(reader.readLine());

            String[] fields = reader.readLine().split("\t");
            Card[] reserve = new Card[RESERVE_SIZE];
            for (int i=0; i<RESERVE_SIZE; i++)
                reserve[i] = Card.valueOf(Integer.parseInt(fields[i]));
            board.setReserve(reserve);

            for (int c=0; c<NCOLS; c++) {
                for (String field : reader.readLine().split("\t")) {
                    if (field.length() > 0)
                        board.addCard(c, Card.valueOf(Integer.parseInt(field)));
                }
            }

            int deal = 0;
            long time = 0;
            List<Step> steps = new ArrayList<Step>();
            String line;
            while ((line = reader.readLine()) != null) {
                fields = line.split("\t");
                if (fields[0].equals("deal")) {
                    deal = Integer.parseInt(fields[1]);
                } else if (fields[0].equals("time")) {
                    time = Long.parseLong(fields[1]);
                } else if (fields[0].equals("steps")) {
                    for (int i=1; i<fields.length; i++)
                        steps.add(readStep(fields[i]));
                }
            }
            game.start(rule, deal);
            game.setElapsedMillis(time);
            game.restoreHistory(steps);
            return game;
        } catch (RuntimeException e) {
            throw new IOException("Not a saved game", e);
        }
    }

    private static Step readStep(String field) {
        if (field.equals("r"))
            return Step.reserveDeal();
        String[] p = field.split(",");
        Move move = new Move(Integer.parseInt(p[1]), Integer.parseInt(p[2]), Integer.parseInt(p[3]));
        return Step.move(move, Integer.parseInt(p[4]), p[5].equals("1"));
    }
}
